multishard_mutation_query.cc

```diff
@@ -781,6 +782,44 @@ future<foreign_ptr<lw_shared_ptr<typename ResultBuilder::result_type>>> do_query
     co_return f.get0();
 }
 
+template <typename ResultBuilder>
+future<foreign_ptr<lw_shared_ptr<typename ResultBuilder::result_type>>> do_query_tablets(
+        distributed<replica::database>& db,
+        schema_ptr s,
+        const query::read_command& cmd,
+        const dht::partition_range_vector& ranges,
+        tracing::trace_state_ptr trace_state,
+        db::timeout_clock::time_point timeout,
+        noncopyable_function<ResultBuilder()> result_builder_factory) {
+    auto& table = db.local().find_column_family(s);
+    auto erm = table.get_effective_replication_map();
+    const auto& token_metadata = erm->get_token_metadata();
+    const auto& tablets = token_metadata.tablets().get_tablet_map(s->id());
+    const auto this_node_id = token_metadata.get_topology().this_node()->host_id();
+
+    auto query_cmd = cmd;
+    auto result_builder = result_builder_factory();
+
+    std::vector<foreign_ptr<lw_shared_ptr<typename ResultBuilder::result_type>>> results;
+    locator::tablet_range_splitter range_splitter{s, tablets, this_node_id, ranges};
+    while (auto range_opt = range_splitter()) {
+        auto& r = *results.emplace_back(co_await db.invoke_on(range_opt->shard,
+                    [&result_builder, gs = global_schema_ptr(s), &query_cmd, &range_opt, gts = tracing::global_trace_state_ptr(trace_state), timeout] (replica::database& db) {
+            return result_builder.query(db, gs, query_cmd, range_opt->range, gts, timeout);
+        }));
+
+        // Substract result from limit, watch for underflow
+        query_cmd.partition_limit -= std::min(query_cmd.partition_limit, ResultBuilder::get_partition_count(r));
+        query_cmd.set_row_limit(query_cmd.get_row_limit() - std::min(query_cmd.get_row_limit(), ResultBuilder::get_row_count(r)));
+
+        if (!query_cmd.partition_limit || !query_cmd.get_row_limit() || r.is_short_read()) {
+            break;
+        }
+    }
+
+    co_return result_builder.merge(std::move(results));
```

Comment:
Do we, in fact, need to merge? Can't we return an empty page per tablet?

With vnodes, we could have a large number of vnode ranges intersected with a large number of shards-ranges, yielding a huge number of vnode-shard-ranges so we needed to query in parallel etc. But now we have a limited number of tablets per node (up to 10,000, so not trivial), but we also support empty pages. So maybe it's enough to read from the nearest tablet and return a short read if we reach the end.

Reply:
This pushes the loop to the caller (and potentially to the client), but it's already prepared for it due to short reads. This can simplify the code here.

Reply:
In this case the client is internal so we are not saving any complexity here.